### PR TITLE
Auto-update of the documentation buffer

### PR DESCRIPTION
**Pull request by galaunay**
Following feature request #1276.

## Summary
This PR adds the automatic update of the documentation buffer based on the symbol at point.
Documentation is also updated in company, based on the currently selected candidate.

To avoid documentation poping up unexpectedly, the documentation is only updated when the documentation buffer is visible (i.e. associated to a window).

## Control of the feature
The delay before which updating the documentation is controlled by the option `elpy-autodoc-delay`, that can be set to `nil` to deactivate the feature.

### CODE REVIEW

File: elpy.el [hunk 1]
```diff
@@ -593,9 +593,13 @@ virtualenv.
...
    (elpy-mode
-    (elpy-modules-buffer-init))
+    (elpy-modules-buffer-init)
+    (add-hook 'pre-command-hook 'elpy-autodoc--pre-command nil t)
+    (add-hook 'post-command-hook 'elpy-autodoc--post-command nil t))
    ((not elpy-mode)
-    (elpy-modules-buffer-stop))))
+    (elpy-modules-buffer-stop)
+    (remove-hook 'pre-command-hook 'elpy-autodoc--pre-command t)
+    (remove-hook 'post-command-hook 'elpy-autodoc--post-command t))))
```
Comment (jorgenschaefer) [inline elpy.el]: Can't this be a module? Then this init/deinit can be moved to a module definition, and people can deactivate it entirely when they want.

**Comment by Nicholas-Mitchell** [comment]
It was me who originally asked for this feature - big thank to @galaunay for the implementation.
I have tested it with my setup in Emacs 25.1.1 with the latest `elpy` from melpa ():

        Status: Installed in ‘elpy-20180209.146/’ (unsigned).
        Version: 20180209.146

I'm happy to report it works really well, as desired, with my current setup. I use the ipython interpreter and so with the latest elpy version, I just had to update my config from the old `(elpy-use-ipython)` to the new cleaner way:

    (setq python-shell-interpreter "ipython"
        python-shell-interpreter-args "-i --simple-prompt")

I didn't have to change anything.

The 0.5 second refresh rate is perfect - no feeling of lag whatsoever, but I don't have the feeling I have to wait on it to read the docs.

**Comment by galaunay** [comment]
@jorgenschaefer I agree with the module. I am working on it.

@Nicholas-Mitchell, thanks for the backup.
Happy to see that It works well and that you find it useful as it is :).

Just an additional piece of information:
Autodoc is not as smart as the classical `elpy-doc`.
For example in this case (`|` being the cursor):
```Python
res = foo(a|, b, c)
```
`elpy-doc` gives the documentation of `foo` while autodoc give the documentation of the type of `a` (which is not very usefull in general...).

This is linked to the fact that I wanted autodoc to work asynchronously.
I don't see any (simple) way of make autodoc smarter...

**Comment by galaunay** [comment]
Done, should be good.

**Comment by Nicholas-Mitchell** [comment]
@galaunay - just an idea about your last point - autodoc returning the documentation for types etc.
Would it be possible to bootstrap something like paredit, to know what the current most relevant sexp is, then pipe that into elpy-doc asynchronously?

Even with this small issue, I am really enjoying autodoc - thanks again!

**Comment by galaunay** [comment]
Good idea.

Currently, when using `elpy-doc`, elpy looks for documentation at points, and if it is not relevant, move backward with `elpy-nav-backward-up-list` and search again for documentation.

To stay asynchronous, we need to know a priori (without calling for documentation) if the current symbol under point is relevant or not.
The only way I could find to do that is to check if the symbol under point ends with an opening parenthesis...

@jorgenschaefer, do you know if python-mode provide a better way of doing this ?

**Comment by jorgenschaefer** [comment]
"Relevant" is a tricky concept :-D I doubt python-mode can do that. We used to have some attempts to do that, and it did not work well.

For this feature, you could set a marker at point, call the backend, and when it returns with no info, continue from the marker instead of point. Just remember to clean up the marker.

Or remember point and continue only if point has not changed, if that works with this workflow. You run a new async call when point changes anyhow, I think.

**Comment by galaunay** [comment]
Ok, 

I removed the parenthesis thing, that was a little subjective (I agree).
After all, somebody may want to access the documentation of builtins type :).

I will use it as it is, see if it really needs a smarter doc function.

**Comment by galaunay** [comment]
@jorgenschaefer do you agree to make this a default behavior ?

**Comment by jorgenschaefer** [comment]
What do you mean with "this", exactly? The whole module?

**Comment by galaunay** [comment]
Sorry for the delay (I somehow missed your reply).

I meant to activate the autodoc module by default, 
which will imply that the documentation will be updated with the thing under point by default.

I can see several reasons for which users may not want this functionality, that's why I am unsure about activating it by default or not.

**Comment by Nicholas-Mitchell** [comment]
Although I love the functionality, I would suggest leaving it off by
default and set a toggle to activate it. The reason being that beginners
may run into problems when documentation isn't found (due to virtual
environment and path issues) or the RPC process times out. In my opinion,
this would negatively effect the entire Elpy out-of-the-box experience.

The documentation should just make it clear, how one activates it.

**Comment by galaunay** [comment]
I updated the documentation and deactivated the module by default.

This should change nothing to the Elpy default behavior, tests are passing and I just checked it on one of my script to ensure that everything was fine.

Seems ready to merge for me.

Thanks all for your help on this.

**Pull request merged**